src/sentry/auth/access.py

```diff
@@ -196,7 +222,11 @@ def has_project_scope(self, project: Project, scope: str) -> bool:
 
         >>> access.has_project_scope(project, 'project:read')
         """
-        return self.has_project_access(project) and self.has_scope(scope)
+        if not self.has_project_access(project):
+            return False
+        if self.has_scope(scope):
+            return True
+        return any(self.has_team_scope(team, scope) for team in project.teams.all())
```

Comment:
I think this will make a query for all project teams whenever this function is called. It's possible django does some smart caching here, but we'd need to verify. Since this is called in a loop here we might want to make sure it's cached somehow: https://github.com/getsentry/sentry/blob/2464dbd71d1460c332d33881cb1943e345c79480/src/sentry/api/bases/project.py#L26-L33

Reply:
Good catch. I'll investigate that, and I have a few ideas for how to refactor around it.

Reply:
This looks good to me. My only concern here is that in the past these access methods typically don't make IO calls (correct me if I'm wrong there), but just operate on data in the access object itself. Probably not enough of a concern to block the change, but maybe just keep an eye on performance and make sure things don't end up slower due to this.

(partly related, I just noticed that `OrganizationGlobalMembership` does a bunch of IO, and is probably part of the reason that orgs with a lot of projects load so slowly for global org members)

Reply:
> This looks good to me. My only concern here is that in the past these access methods typically don't make IO calls (correct me if I'm wrong there), but just operate on data in the access object itself.

Yep, good point. I don't see any way around querying `project.teams.all()` for this new feature, since we didn't need that information before but we do now, and the Access object doesn't preload project-to-team relationships.

> (partly related, I just noticed that `OrganizationGlobalMembership` does a bunch of IO, and is probably part of the reason that orgs with a lot of projects load so slowly for global org members)

Just making sure, you mean `OrganizationGlobalMembership` and not `OrganizationGlobalAccess`? `OrganizationGlobalMembership` is only from the `_from_sentry_app` flow and it's always preloaded a lot of projects.

`OrganizationGlobalAccess` temporarily had a performance problem but it should not be preloading large numbers of projects as of a few weeks ago. If Access objects for global org members are still loading large sets, it would now be doing it as a `cached_property`, which should not be any slower overall but might move a bottleneck to later in the flow.

> Probably not enough of a concern to block the change, but maybe just keep an eye on performance and make sure things don't end up slower due to this.

Definitely. If we can isolate a problem, I'd bet we can attack it by factoring out dependency on the `teams` and `projects` attributes. I'd make them private if I could.

Reply:
> Just making sure, you mean OrganizationGlobalMembership and not OrganizationGlobalAccess? OrganizationGlobalMembership is only from the _from_sentry_app flow and it's always preloaded a lot of projects.

Didn't notice there were two of these. Nbd then, although they are a bit surprisingly named.

Reply:
Yeah, sorry about the confusing names. Context is here: #32967